src/trans/gpu/internal/trgtol_mod.F90

```diff
@@ -8,1366 +10,618 @@
-  LOGICAL   :: LLUV(KF_GP),LLGP2(KF_GP),LLGP3A(KF_GP),LLGP3B(KF_GP)
-  LOGICAL :: LLDONE, LLINDER
-  INTEGER(KIND=JPIM) :: IUVLEVS(KF_GP),IUVPARS(KF_GP),IGP2PARS(KF_GP)
-  INTEGER(KIND=JPIM) :: IGP3APARS(KF_GP),IGP3ALEVS(KF_GP),IGP3BPARS(KF_GP),IGP3BLEVS(KF_GP)
-  INTEGER(KIND=JPIM) :: IUVPAR,IUVLEV,IGP2PAR,IGP3ALEV,IGP3APAR,IGP3BLEV,IGP3BPAR,IPAROFF
-  INTEGER(KIND=JPIM) :: IOFF,IOFF1,IOFFNS,IOFFEW,J1,J2
-  INTEGER(KIND=JPIM) :: KINDEX(D%NLENGTF),INDOFF(NPROC),IFLDOFF(MAX(KF_GP,KF_FS))
-  INTEGER(KIND=JPIM) :: ISEND_FLD_TOTAL(NPROC),ISEND_FLD_END
-  INTEGER(KIND=JPIM) :: IRECV_FLD_END
-  INTEGER(KIND=JPIM) :: INUMFLDS
-  INTEGER(KIND=JPIM) :: IGPTRSEND(2,NGPBLKS,NPRTRNS)
-  INTEGER(KIND=JPIM) :: IGPTRRECV(NPRTRNS)
-  INTEGER(KIND=JPIM) :: IGPTROFF(NGPBLKS)
-  !     INTEGER FUNCTIONS
-  INTEGER(KIND=JPIM) :: IRECVCOUNT,ISENDCOUNT
-  INTEGER(KIND=JPIM) :: IJPOS(NGPBLKS),IFLDA(KF_GP),JJ,JI,IFLDT
-  REAL(KIND=JPHOOK) :: ZHOOK_HANDLE
-  REAL(KIND=JPHOOK) :: ZHOOK_HANDLE_BAR
-  
-  INTEGER(KIND=JPIM) :: IERROR, irank
-  
-  REAL(KIND=JPRBT) :: TIMEF, tc
-  
-  #ifdef PARKINDTRANS_SINGLE
-  #define TRGTOL_DTYPE MPI_REAL
-  #else
-  #define TRGTOL_DTYPE MPI_DOUBLE_PRECISION
-  #endif
-
-  !     ------------------------------------------------------------------
-  
-  !*       0.    Some initializations
-  !              --------------------
-  
-  IF (LHOOK) CALL DR_HOOK('TRGTOL_CUDAAWARE',0,ZHOOK_HANDLE)
-  
-  iunit=300+myproc
-
-  CALL GSTATS(1805,0)
-  
-  LLINDER = .FALSE.
-  LLPGPUV = .FALSE.
-  LLPGP3A = .FALSE.
-  LLPGP3B = .FALSE.
-  LLPGP2  = .FALSE.
-  LLPGPONLY = .FALSE.
-  IF(PRESENT(KPTRGP))  LLINDER = .TRUE.
-  IF(PRESENT(PGP))     LLPGPONLY = .TRUE.
-  IF(PRESENT(PGPUV))   LLPGPUV = .TRUE.
-  IF(PRESENT(PGP3A))   LLPGP3A = .TRUE.
-  IF(PRESENT(PGP3B))   LLPGP3B = .TRUE.
-  IF(PRESENT(PGP2))    LLPGP2 = .TRUE.
-  IUVPAR=0
-  IUVLEV=0
-  IOFF1=0
-  IOFFNS=KF_SCALARS_G
-  IOFFEW=2*KF_SCALARS_G
-  LLUV(:) = .FALSE.
-  IUVPARS(:) = -99
-  IUVLEVS(:) = -99
-  IF (LLPGPUV) THEN
+
+  PRIVATE
+  PUBLIC :: TRGTOL_HANDLE, TRGTOL, PREPARE_TRGTOL
+
+  TYPE TRGTOL_HANDLE
+    TYPE(ALLOCATION_RESERVATION_HANDLE) :: HCOMBUFS, HCOMBUFR_AND_REEL
+  END TYPE
+CONTAINS
+  FUNCTION PREPARE_TRGTOL(ALLOCATOR,KF_GP,KF_FS) RESULT(HTRGTOL)
+    USE PARKIND_ECTRANS, ONLY : JPIM, JPRB, JPRBT
+    USE ALLOCATOR_MOD
+    USE TPM_DISTR, ONLY: D
+
+    IMPLICIT NONE
+
+    TYPE(BUFFERED_ALLOCATOR), INTENT(INOUT) :: ALLOCATOR
+    INTEGER(KIND=JPIM), INTENT(IN) :: KF_GP, KF_FS
+    TYPE(TRGTOL_HANDLE) :: HTRGTOL
+
+    REAL(KIND=JPRBT) :: DUMMY
+
+    INTEGER(KIND=C_SIZE_T) :: NELEM
+
+    HTRGTOL%HCOMBUFS = RESERVE(ALLOCATOR, KF_GP*D%NGPTOT*SIZEOF(DUMMY))
+
+    NELEM = KF_FS*D%NLENGTF*SIZEOF(DUMMY) ! ZCOMBUFR
+    NELEM = NELEM + KF_FS*D%NLENGTF*SIZEOF(DUMMY) ! PREEL_REAL
+    HTRGTOL%HCOMBUFR_AND_REEL = RESERVE(ALLOCATOR, NELEM)
+  END FUNCTION
+
+  SUBROUTINE TRGTOL(ALLOCATOR,HTRGTOL,PREEL_REAL,KF_FS,KF_GP,KF_UV_G,KF_SCALARS_G,&
```

Comment:
I dropped the non-cuda aware path and moved everything to OpenACC here.  This is pretty much a rewrite of this function. Note that the extra routines (inigptr) have been dropped and integrated here. I gave my best to properly comment the code because this memory layouts are really tough to undersatnd. Also all codes in TRLTOG/TRGTOL are made sure to be very clearly "reversed" each other.